### ISSUE TITLE
protected override void Execute(RuleContext context) is executed for the 1st child node only not the next ones

### ISSUE THREAD

**Issue by Umplify** [MarimerLLC/cslaforum#296]
Hello, Please take a look at the code snippet below:
```
FieldManager.GetChildren()?.ForEach(child => 
            {
                if(child is IEditableCollection)
                {
                    foreach(var grandChild in (IEnumerable<IRuleEnforcement>)child)
                    {
                        grandChild.CheckBusinessRules();
                    }
                }
            });
```
Each child has at least two grandChild nodes. When grandChild.CheckBusinessRules() is invoked for the first time, 'protected override void Execute(RuleContext context)' is invoked too. The 2nd iteration of the 'foreach' loop for the 2nd grandChild is expected to invoke 'protected override void Execute(RuleContext context)' again but it's not actually happening! As a side note, grandChild.CheckBusinessRules() is invoking CSLA's 'BusinessRules.CheckRules()'. Any clue why the 2nd child's "protected override void Execute(RuleContext context)" method is not called despite the 1st child in the node is valid and there is no hindrance to proceed with validating the 2nd child?

**Comment by jonnybee** [comment]
Are you able to create a small sample project that recreates this behaviour that you can share with us? 

There is logic in the RuleEngine to process rules per-property in a given order (Priority ASC) and a property that says ProcessThroughPriority (default value =0). I have seen several times where developers have misunderstood the execution sequence and used a Priority on an "object" level - but rules are always executed on e per-property 

Rules is processed in the order of Priority and if any rule shortcircuits then the remaining rules will not be processed for that property. If no rules have shortcircuited the RuleEngine will process all rules where Priority is less than or equal to ProcessThroughPriority. If any rules have RuleSeverity.Error the rule engine will stop executing rules with a "lower" Priority.

So - depending on how you rules have been coded and assembled - this could be  the expected behaviour. 

So - we need to see more of your code to understand your problem.

**Comment by Umplify** [comment]
@jonnybee Thanks Jonny. I will review the code to ensure that the properties' priorities are properly set. I will provide you the requested source code similar to what I have if my investigation is not conclusive.

**Comment by Umplify** [comment]
@jonnybee I created a new project _(attached to this thread)_ with simplified objects whose architect is derived from my main project and 100% similar. Here are a few points to consider:

1. Iterating through children in both main and the new project is the same.
2. Both main project's and the new project's parent objects contain two list objects each and each list has two nodes/granchild only.
3. The main project's validation rules' `protected override void Execute(RuleContext context)` is executed against the first grandchild only and it is never executed against the next grand child under the same list object despite the first grand child is valid. But, `protected override void Execute(RuleContext context)` is executed against both grandchildren under the very same list object in the attached new project!
4. During debugging the attached new project, I realized that ProcessThroughPriority is always zero despite I initialized it to 1 within "AddBusinessRules" methods in Foo and Zoo objects.
5. During debugging the **main project**, I realized that ProcessThroughPriority is '1' before validating the first grandchild object but it's reset to zero when iterating to the 2nd grandchild of the same type under the very same list object. I attached two screenshots to this thread to depict this scenario.

Bottom line, I cannot justify why both grandchild nodes' validation are executed in the new project but not in the main project despite the structure of objects are the same and also the way that objects are deemed as valid are alike too. Also, I cannot justify why ProcessThroughPriority does not stay consistent with its initialization.

[RuleValidationTest.zip](https://github.com/MarimerLLC/cslaforum/files/696928/RuleValidationTest.zip)

![screenshot1](https://cloud.githubusercontent.com/assets/4282484/21818329/5ae4d096-d735-11e6-9ebc-4cd9028d5fed.png)
![screenshot2](https://cloud.githubusercontent.com/assets/4282484/21818328/5ae23f70-d735-11e6-879e-6eb9f9bbf6ec.png)

**Comment by jonnybee** [comment]
Well, I cannot explain that either but something must be different. 

Also - there may be some understanding about how the rule engine works. 
This makes no sense: 

            BusinessRules.AddRule(new Dependency(MinValueProperty, MaxValueProperty) { Priority = 0 });
            BusinessRules.AddRule(new Dependency(MaxValueProperty, MinValueProperty) { Priority = 0 });
            BusinessRules.AddRule(new FooValidRules() { Priority = 1 });

As there is no property rule on either MinValue/MaxValue so there is no point in declaring a Dependency.
Also the FooValidRules is an object level rule so it will never be called as the user edits the objects field in a rich client. 

ProcessThroughPriority is NOT an issue in this example and you should add the object level rule with priority=0

**Comment by Umplify** [comment]
Thanks Jonny. I substituted object level rules with property level rules and made them all have `Priority = 0` in both main project and the attached sample project. The issue of not executing validation against the rest of grand child objects still persists:

- The validation rule is executed against both two Foo and both two Zoo objects (deemed as grandchildren) in the attached sample project.
- The validation rule is executed only against the 1st Foo-like and the 1st Zoo-like object in the **main project** and the rest of objects are not validated despite the first object is always valid. 

I reviewed main project's code to be sure that the way that validation rules are assembled are similar to the attached sample project. I have to say that there is no difference in that context. At this point I am quite stuck with making the main project work.
[RuleValidationTest.zip](https://github.com/MarimerLLC/cslaforum/files/699176/RuleValidationTest.zip)

**Comment by Umplify** [comment]
@jonnybee I found the root cause of the issue after reviewing both main project and the sample project. I missed a piece of code to include in BaseObject.cs which is according to the code snippet below:

```
        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.RuleSet = "DefaultRuleSet";
        }
```

In this code snippet, "DefaultRuleSet" is my own default rule set name due to a business reason.

After adding this piece of code to the base object, the issue started manifesting itself and it can be duplicated in the attached project. 

Basically, all Foo and Zoo object instances will receive the same rule set name i.e. "DefaultRuleSet" when base class's method is called. When the debugger hit the line below, I examined BusinessRule.RuleSet property for each child returned by FileManager.GetChildren() method and none of them possessed my rule set named "DefaultRuleSet" and all reset to CSLA's default rule set name i.e. **default**. (please see the screenshot)

```
            var children = FieldManager.GetChildren()?
                .Where(child => child is IEditableCollection)?
                .ToList();
```

I couldn't find anything in the documentation to justify this behavior. If you execute the attached code (different from the previous ones), you will realize that business rules for each property are not invoked.

Is that an issue / bug to address in CSLA or it's an improper configuration in assembling rules?

Thanks!

Source code to download --> [RuleValidationTest.zip](https://github.com/MarimerLLC/cslaforum/files/702197/RuleValidationTest.zip)

![capture](https://cloud.githubusercontent.com/assets/4282484/21897285/25c53710-d8b7-11e6-8931-5264e2dcebce.PNG)

**Comment by jonnybee** [comment]
There is 2 important mistakes in your code. 

1. Make sure to set your own custom RuleSetName -before_ any rules is added. This is ONLY for registering rules into your own custom rule set name (and I would rather encourage you to NOT code like this - see 2). Remember: AddBusinessRules is only called ONCE per type and base.AddBusinessRules may also add business rules (or DataAnnotations). 

        protected override void AddBusinessRules()
        {
            BusinessRules.RuleSet = "DefaultRuleSet";
            base.AddBusinessRules();
        }

2. Typically in you create/fetch method your code must set which ruleset to use for this instance. If none is set (or. failure to set ruleset name) the rules in CSLA default name is executed (and here there are no rules in you code). 

        private void Child_Fetch(ZooObject zoo)
        {
            MinValue = zoo.MinValue;
            MaxValue = zoo.MaxValue;
            // Set which ruleset top use for this instance
            BusinessRules.RuleSet = "DefaultRuleSet";
        }

        private void Child_Fetch(FooObject foo)
        {
            MinValue = foo.MinValue;
            MaxValue = foo.MaxValue;
            // Set which ruleset top use for this instance
            BusinessRules.RuleSet = "DefaultRuleSet";
        }

The option for having multiple rulesets is that your code must set which ruleset to use when the object i Created or Fetched. if this is not the default in CSLA.

**Comment by Umplify** [comment]
@jonnybee  would it be possible to reflect these points to the next edition of CSLA documentations?

**Comment by jonnybee** [comment]
In your case - as AddBusinessRules is called only once this is the only object that has the correct ruleset set. 

It is recommended to have the lines at the end og AddBusinessRules: 

    // use default rule set
    BusinessRules.RuleSet = ApplicationContext.DefaultRuleSet;

This is documented in the Using CSLA 4  - Creating Business Object ebook.

**Comment by jonnybee** [comment]
It is also documented in this blog post
http://www.lhotka.net/weblog/CSLA4BusinessRulesSubsystem.aspx

**Comment by Umplify** [comment]
@jonnybee  thanks Jonny. I believe that here is the key statement:

> The RuleSet value is serialized along with the state of the business object, so you can (and typically will) set the RuleSet property in your DataPortal_Create() and DataPortal_Fetch()

**Comment by Umplify** [comment]
@jonnybee Many thanks for the support. I adjusted the code and it works fine now. Closing this issue.
